src/sqlfluff/utils/reflow/reindent.py

```diff
@@ -27,3 +36,792 @@ def deduce_line_indent(raw_segment: RawSegment, root_segment: BaseSegment) -> st
+            # or placeholders for consumed whitespace). This is a bug.
+            raise NotImplementedError(
+                "Unexpected templated indent. Report this as a bug on GitHub."
+            )
+    else:
+        current_indent = ""
+
+    # First handle starting indent.
+    assert "\n" not in current_indent, f"Found newline in indent: {indent_seg}"
+    desired_indent_units = indent_line.desired_indent_units(forced_indents)
+    desired_starting_indent = desired_indent_units * single_indent
+    initial_point = cast(ReflowPoint, elements[indent_points[0].idx])
+
+    if current_indent != desired_starting_indent:
+        reflow_logger.debug(
+            "    Correcting indent @ line %s. Existing indent: %r -> %r",
+            elements[indent_points[0].idx + 1].segments[0].pos_marker.working_line_no,
+            current_indent,
+            desired_starting_indent,
+        )
+        # Initial point gets special handling it it has no newlines.
+        if indent_points[0].idx == 0 and not indent_points[0].is_line_break:
+            new_fixes = [
+                LintFix.delete(seg, description="First line should not be indented.")
+                for seg in initial_point.segments
+            ]
+            new_point = ReflowPoint(())
+        # Placeholder indents also get special treatment
+        else:
+            new_fixes, new_point = initial_point.indent_to(
+                desired_starting_indent,
+                **anchor,  # type: ignore
+            )
+        elements[indent_points[0].idx] = new_point
+        fixes += new_fixes
+
+    closing_balance = indent_points[-1].closing_indent_balance
+    # Edge case for first line of a file (where starting indent must be zero).
+    if indent_points[-1].last_line_break_idx is None:
+        starting_balance = 0
+    else:
+        starting_balance = indent_points[0].closing_indent_balance
+
+    # Then check for new lines. Either on the way up...
+    if closing_balance > starting_balance:
+        if indent_points[-1].indent_trough:
+            closing_trough = (
+                indent_points[-1].initial_indent_balance
+                + indent_points[-1].indent_trough
+            )
+        else:
+            closing_trough = (
+                indent_points[-1].initial_indent_balance
+                + indent_points[-1].indent_impulse
+            )
+
+        # On the way up we're looking for whether the ending balance
+        # was an untaken indent on the way up.
+        if (
+            # Edge case: if closing_balance > starting balance
+            # but closing_trough isn't, then we shouldn't insert
+            # a new line. That means we just dropped back down to
+            # close the untaken newline.
+            closing_trough > starting_balance
+            and closing_trough in indent_points[-1].untaken_indents
+        ):
+            # It was! Force a new indent there.
```

Comment:
Given the "if" comment is pretty complex, I'm unsure what exactly is meant by "It was" here.

Reply:
I'll rework all the comments in this area.